nixos/gitlab: Clean config dir, switch to puma, split PreStart script and more

talyz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Various improvements to the GitLab service:

#### Clean up the config dir more thoroughly

Remove all the subdirectories in `config` on start.

From one version of GitLab to the next, the files in the `config` directory changes. Since we're only overwriting the existing files
with ones from the repo, cruft sometimes gets left behind, occationally causing issues.

Ideally, all configuration put in the `config` directory is declared by NixOS options and we could just remove the whole directory on start, but I'm not sure if that's the case. It would also require a little bit of additional rework and testing. The subdirectories,
however, should seldom contain user configuration and the ones that frequently does, `initializers`, is already removed on start.


#### Switch from unicorn to puma

Puma is the new upstream default server since GitLab 13.

#### Make gitlab.service's `PreStart` into two new services

Make the config initialization script run in gitlab.service's `PreStart` section into two new services, `gitlab-config.service` and
`gitlab-db-config.service`. Other services can then depend on the config scripts they need instead of unnecessarily depending on
`gitlab.service`. This makes the reason for the configured service dependencies much clearer and should also reduce the restart time of the `gitlab` service quite a lot, when triggered manually.

Also, set up stricter service dependencies, using `bindsTo`, to ensure that if a service fails or is stopped, its dependants are also
stopped. For example, if we're using the `postgresql` service and it's stopped, `gitlab.service` and `gitlab-sidekiq.service`, which depend on it to function, should also be stopped.


#### Change default SMTP port, enable postfix only if used

Change the default SMTP port to `25`, to better match the default address `localhost`. This gets rid of some error outputs in the test, where it fails to connect to localhost:465.

Also, don't enable postfix by default unless it's actually useful to us.

#### Test pages and mailroom service start

Make sure that the `gitlab-pages` and `gitlab-mailroom` services at least start.

#### Make PSQL a function rather than a variable in gitlab-postgresql



cc @jslight90 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

talyz committed:
nixos/gitlab: Clean up the config dir more thoroughly

This removes all the subdirectories in `config` on start.

From one version of GitLab to the next, the files in the `config`
directory changes. Since we're only overwriting the existing files
with ones from the repo, cruft sometimes gets left behind,
occationally causing issues.

Ideally, all configuration put in the `config` directory is declared
by NixOS options and we could just remove the whole directory on
start, but I'm not sure if that's the case. It would also require a
little bit of additional rework and testing. The subdirectories,
however, should seldom contain user configuration and the ones that
frequently does, `initializers`, is already removed on start.

talyz committed:
nixos/gitlab: Switch from unicorn to puma

Puma is the new upstream default server since GitLab 13.

talyz committed:
nixos/gitlab: Change default SMTP port, enable postfix only if used

Change the default SMTP port to `25`, to better match the default
address `localhost`. This gets rid of some error outputs in the test,
where it fails to connect to localhost:465.

Also, don't enable postfix by default unless it's actually useful to
us.

talyz committed:
nixos/gitlab: postgresql: Make PSQL a function, run as superUser

A function is more appropriate for this use. See
http://mywiki.wooledge.org/BashFAQ/050 for reference.

Also, we don't need to run the service as root: since we essentially
run all commands as `services.postgresql.superUser` anyway, the whole
service can just run as that user instead.

talyz committed:
nixos/gitlab: Make gitlab.service's PreStart into two new services

Make the config initialization script run in gitlab.service's PreStart
section into two new services, `gitlab-config.service` and
`gitlab-db-config.service`. Other services can then depend on the
config scripts they need instead of unnecessarily depending on
`gitlab.service`. This makes the reason for the configured service
dependencies much clearer and should also reduce the restart time of
the `gitlab` service quite a lot, when triggered manually.

Also, set up stricter service dependencies, using `bindsTo`, to ensure
that if a service fails or is stopped, its dependants are also
stopped. For example, if we're using the `postgresql` service and it's
stopped, `gitlab.service` and `gitlab-sidekiq.service`, which depend on
it to function, should also be stopped.

talyz committed:
nixosTests.gitlab: Test pages and mailroom service start

Make sure that the `gitlab-pages` and `gitlab-mailroom` services at
least start.

aanderse commented on nixos/modules/services/misc/gitlab.nix:
```diff
@@ -655,36 +655,43 @@ in {
 
-        PSQL="${pkgs.util-linux}/bin/runuser -u ${pgsql.superUser} -- psql --port=${toString pgsql.port}"
+        PSQL() {
+            runuser -u ${pgsql.superUser} -- psql --port=${toString pgsql.port} "$@"
```
This doesn't need to run as `root` anymore. Can set `User` to `pgsql.superUser` and should be fine.

talyz replied:
Good point! It's fixed now.

aanderse reviewed (commented):
Justa few minor things I noticed. I didn't review the entire change in detail but I think this is a great change. Nice job!

dasJ commented:
Nice work @talyz!
I have the following changes running successfully in production:
- config dir cleanup improvement
- Switch from unicorn to puma

Everything works just fine. The requested postgres change seems reasonable but I cannot really test it since I don't use that functionality of the module

talyz commented on nixos/modules/services/misc/gitlab.nix:
```diff
@@ -424,7 +424,7 @@ in {
         port = mkOption {
           type = types.int;
-          default = 465;
+          default = 25;
```
Yeah, you're right that it could. I've added a release note. :)

cole-h commented:
@ofborg eval

talyz committed:
nixos/gitlab: Add `gitlab.target`

To make it easier to start and stop all GitLab services, introduce
`gitlab.target` which wants all services (meaning they will start with
it) and which all services are part of (meaning they will stop with
it).

talyz commented:
I've pushed a commit that adds a target for all GitLab services, to make it easier to start and stop them.

SuperSandro2000 commented on nixos/modules/services/misc/gitlab.nix:
```diff
@@ -726,13 +739,156 @@ in {
+            set -eu
+
+            chown --no-dereference '${cfg.user}':'${cfg.group}' '${cfg.statePath}'/*
+            if [[ ! -z "$(ls -A '${cfg.statePath}'/config/)" ]]; then
```
```suggestion
            if [[ -n "$(ls -A '${cfg.statePath}'/config/)" ]]; then
```
